Mix-blend-mode incorrectly blends with view base color.

mattwoodrow opened the pull request:
#### 4f26343f6d0bd701a726e1967b7d65cd24f98d51
<pre>
Mix-blend-mode incorrectly blends with view base color.
<a href="https://bugs.webkit.org/show_bug.cgi?id=261708">https://bugs.webkit.org/show_bug.cgi?id=261708</a>
&lt;rdar://problem/115688282&gt;

Reviewed by Simon Fraser.

Mix-blend-mode on an element that blends into the &apos;Root Element Group&apos; should have only the root element&apos;s background color, or transparent if none. We&apos;re including the opaque white base color of the RenderView too, which is incorrect.

This change changes BackgroundPainter so that we only draw the base color if it&apos;s ok to draw it as part of the root element background (if the root background isn&apos;t an isolated group).

It also changes RenderView to make sure we do paint the base color (for iframes, particularly) if the root element won&apos;t be painting it.

Removes some tests (including one WPT) that were explicitly trying to test the wrong thing (and failed on everything except WebKit).

* LayoutTests/css3/blending/blend-mode-body-child-expected.html: Removed.
* LayoutTests/css3/blending/blend-mode-body-child.html: Removed.
* LayoutTests/css3/blending/blend-mode-body-composited-child-expected.html: Removed.
* LayoutTests/css3/blending/blend-mode-body-composited-child.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/compositing_simple_div-expected.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/compositing_simple_div-ref.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/compositing_simple_div.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/mix-blend-mode/mix-blend-mode-root-element-group-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/mix-blend-mode/mix-blend-mode-root-element-group-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/mix-blend-mode/mix-blend-mode-root-element-group.html: Added.
* LayoutTests/svg/css/mix-blend-mode-with-inline-svg.html:
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer):
* Source/WebCore/rendering/RenderLayerInlines.h:
(WebCore::RenderLayer::hasNonOpacityTransparency const):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::paintBoxDecorations):
(WebCore::RenderView::rootElementShouldPaintBaseBackground const):
* Source/WebCore/rendering/RenderView.h:

Canonical link: <a href="https://commits.webkit.org/268173@main">https://commits.webkit.org/268173@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c173f901975991607162b5ea58d62c82d96e7425

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18898 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19239 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19845 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20766 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17674 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19088 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22538 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19369 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19465 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19123 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19242 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/16457 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21650 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16457 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17222 "6 flakes 1 failures") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17491 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17394 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21569 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17977 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17052 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/17056 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4494 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21415 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17813 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderView.cpp:
```diff
@@ -693,6 +693,18 @@ bool RenderView::shouldPaintBaseBackground() const
+
+bool RenderView::rootElementShouldPaintBaseBackground() const
+{
+    Element* documentElement = document().documentElement();
```
`auto`

smfr commented on Source/WebCore/rendering/RenderView.cpp:
```diff
@@ -693,6 +693,18 @@ bool RenderView::shouldPaintBaseBackground() const
+    Element* documentElement = document().documentElement();
+    if (RenderElement* rootRenderer = documentElement ? documentElement->renderer() : nullptr) {
+        // The document element's renderer is currently forced to be a block, but may not always be.
+        auto* rootBox = dynamicDowncast<RenderBox>(*rootRenderer);
```
Odd to use `dynamicDowncast<>` but not then null-check

mattwoodrow committed:
Mix-blend-mode incorrectly blends with view base color.
https://bugs.webkit.org/show_bug.cgi?id=261708
<rdar://problem/115688282>

Reviewed by Simon Fraser.

Mix-blend-mode on an element that blends into the 'Root Element Group' should have only the root element's background color, or transparent if none. We're including the opaque white base color of the RenderView too, which is incorrect.

This change changes BackgroundPainter so that we only draw the base color if it's ok to draw it as part of the root element background (if the root background isn't an isolated group).

It also changes RenderView to make sure we do paint the base color (for iframes, particularly) if the root element won't be painting it.

Removes some tests (including one WPT) that were explicitly trying to test the wrong thing (and failed on everything except WebKit).

* LayoutTests/css3/blending/blend-mode-body-child-expected.html: Removed.
* LayoutTests/css3/blending/blend-mode-body-child.html: Removed.
* LayoutTests/css3/blending/blend-mode-body-composited-child-expected.html: Removed.
* LayoutTests/css3/blending/blend-mode-body-composited-child.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/compositing_simple_div-expected.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/compositing_simple_div-ref.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/compositing_simple_div.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/mix-blend-mode/mix-blend-mode-root-element-group-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/mix-blend-mode/mix-blend-mode-root-element-group-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/mix-blend-mode/mix-blend-mode-root-element-group.html: Added.
* LayoutTests/svg/css/mix-blend-mode-with-inline-svg.html:
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer):
* Source/WebCore/rendering/RenderLayerInlines.h:
(WebCore::RenderLayer::hasNonOpacityTransparency const):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::paintBoxDecorations):
(WebCore::RenderView::rootElementShouldPaintBaseBackground const):
* Source/WebCore/rendering/RenderView.h:

Canonical link: https://commits.webkit.org/268173@main
